https://issues.redhat.com/browse/ACM-14631 (#7047)
open-cluster-management/rhacm-docs: gitops/gitops_registering.adoc

. Import one or more managed clusters.

. To register managed clusters to {gitops-short}, complete the link:../clusters/cluster_lifecycle/create_clusterset.adoc#creating-a-managedclusterset[Creating a _ManagedClusterSet_] documentation.

[#register-gitops]
== Registering managed clusters to {gitops}

Complete the following steps to register managed clusters to {gitops-short}:

. Create a managed cluster set binding to the namespace where {gitops-short} is deployed. For an example of binding the managed cluster to the `openshift-gitops` namespace, see the link:https://github.com/stolostron/multicloud-integrations/blob/main/examples/managedclustersetbinding.yaml[`multicloud-integrations`] managed `clusterset` binding example. In the _Additional resources_ section, see _Creating a ManagedClusterSetBinding resource_ for more general information about creating a `ManagedClusterSetBinding`. See link:../clusters/cluster_lifecycle/placement_filter.adoc[Filtering ManagedClusters from ManagedClusterSets] for placement information. 

> . In the namespace that is used in managed cluster set binding, create a `Placement` custom resource to select a set of managed clusters to register to an {ocp-short} {gitops-short} operator instance. Use the `multicloud-integration` placement example as a template. See _Using ManagedClusterSets with Placement_ for placement information. 

Reviewer: The ocp conref isn't required here since the value for {gitops} is Red Hat OpenShift GitOps
```suggestion
. In the namespace that is used in managed cluster set binding, create a `Placement` custom resource to select a set of managed clusters to register to an {gitops-short} operator instance. Use the `multicloud-integration` placement example as a template. See _Using ManagedClusterSets with Placement_ for placement information. 
```

+
*Notes:* 
+

> - Only {ocp-short} clusters are registered to an {ocp} {gitops-short}  operator instance, not other Kubernetes clusters.

Reviewer: ```suggestion
- Only {ocp-short} clusters are registered to an {gitops-short} operator instance, not other Kubernetes clusters.
```

Complete the following steps to register managed clusters to GitOps:

> See the link:../clusters/cluster_lifecycle/create_clusterset.adoc#creating-a-managedclusterset[Creating a _ManagedClusterSet_] documentation for more information.

Reviewer: Can you check if this should be a prereq? The link interrupts the steps.

i know this isn't apart of the changes you made, but will you also update other instances of GitOps by using the conref? I see OpenShift GitOps in the title and on line 4. Please also update this statement: "To configure GitOps with the Push model, you can register a set of one or more {acm} managed clusters to an instance of {ocp} GitOps operator." 

TO

To configure {gitops} with the Push model, you can register a set of one or more {acm} managed clusters to an instance of {gitops-short} operator.

    apiVersion: cluster.open-cluster-management.io/v1beta1
    name: all-openshift-clusters <1>
----
<1> The `placementRef.name` value is `all-openshift-clusters`, and is specified as target clusters for the {gitops-short} instance that is installed in `argoNamespace: openshift-gitops`. The `argoServer.cluster` specification requires the `local-cluster` value.

. Save your changes. You can now follow the {gitops-short} workflow to manage your applications.

[#register-non-ocp]
== Registering non-{ocp-short} clusters to {gitops}

> You can now use the {acm-short} `GitOpsCluster` to register a non-{ocp-short} cluster to a {gitops-short} cluster. With this capability, you can deploy application resources to the non-{ocp-short} cluster by using the {gitops-short} console. To register a non-{ocp-short} cluster to a {gitops-short} cluster, complete the following steps:

Reviewer: ```suggestion
You can now use the {acm-short} `GitOpsCluster` resource to register a non-{ocp-short} cluster to a {gitops-short} cluster. With this capability, you can deploy application resources to the non-{ocp-short} cluster by using the {gitops-short} console. To register a non-{ocp-short} cluster to a {gitops-short} cluster, complete the following steps:
```

----

. To verify that the cluster secret is generated, go to the `openshift-gitops` namespace and confirm that that `GitOpsCluster` resource status reports, `successful`. 
+
*Notes:* 

* With {acm-short} 2.12 or later, the API server URL for all types of non-{ocp-short} `ManagedCluster` resources renders automatically if you use the following importing modes:
** Entering your server URL and API token for the existing cluster. 
** Entering the `kubeconfig` file for the existing cluster. 
* The following cases can make the API server URL empty for one of the `ManagedClusters` resources:
** The non-{ocp-short} cluster is imported to the {acm-short} hub cluster before version 2.12. 
** The non-{ocp-short} cluster is manually imported to the {acm-short} hub cluster in version 2.12 through the import mode, `Run import commands`. 

[#secret-gitops]
== {gitops} token

When you integrate with the {gitops-short} operator, for every managed cluster that is bound to the {gitops-short} namespace through the placement and `ManagedClusterSetBinding` custom resources, a secret with a token to access the `ManagedCluster` is created in the {gitops-short} instance server namespace. 

> The {gitops-short} controller needs this secret to sync resources to the managed cluster. By default, the service account application manager works with the cluster admin permissions on the managed cluster to generate the {gitops-short}  cluster secret in the {gitops-short} instance server namespace. The default namespace is `openshift-gitops`. 

Reviewer: ```suggestion
The {gitops-short} controller needs this secret to sync resources to the managed cluster. By default, the service account application manager works with the cluster administrator permissions on the managed cluster to generate the {gitops-short} cluster secret in the {gitops-short} instance server namespace. The default namespace is `openshift-gitops`. 
```

. Save your changes. You can now follow the GitOps workflow to manage your applications.

[#register-non-ocp]

> == Registering non-_{ocp-short}_ clusters to _{gitops}_

Reviewer: Why is the gitops conref italicized?

Author: I thought we supposed to use _ _ instead of brackets in titles, lol 

Anyway, I did that to be consistent with the {ocp-short} conref that's italicized, I can remove it :)

Reviewer: The brackets are required for the variable to work in the product doc

You can now use the {acm-short} `GitOpsCluster` to register a non-{ocp-short} cluster to a {gitops-short} cluster. With this capability, you can deploy application resources to the non-{ocp-short} cluster by using the {gitops-short} console. To register a non-{ocp-short} cluster to a {gitops-short} cluster, complete the following steps:

> . Go to the API server URL in the non-{ocp-short} `ManagedCluster`  resource `spec` and validate it by running the following command:

Reviewer: there is an extra space before "resource"

+
[source,bash]
----
oc get managedclusters eks-1
----

> . Verify that your output resembles the following:

Reviewer: ```suggestion
. Verify that your output resembles the following information:
```

+
[source,bash]
----
NAME    HUB ACCEPTED   MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
eks-1   true           https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
----

. If the API server URL in the non-{ocp-short} `MangedCluster` resource `spec` is empty, update it manually by completing the following steps: 
.. To complete the API server URL, edit the `MangedCluster` resource `spec` by running the following command: 

+
[source,bash]
----
oc edit managedclusters eks-1
----

> .. Verify that your output resembles the following: 

Reviewer: ```suggestion
.. Verify that your YAML resembles the following file: 
```

+
[source,bash]
----
spec:
  managedClusterClientConfigs:
  - caBundle: ZW1wdHlDQWJ1bmRsZQo=          
    url: https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com
----

.. Save the changes then validate that the API server is completed by running the following command: 

+
[source,bash]
----
oc get managedclusters eks-1
----

> .. Verify that your output resembles the following: 

Reviewer: ```suggestion
.. Verify that your output resembles the following information: 
```

+
[source,bash]
----
NAME     HUB ACCEPTED  MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
eks-1    true          https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
----

. To verify that the cluster secret is generated, go to the `openshift-gitops` namespace and confirm that that `GitOpsCluster` resource status reports, `successful`. 
+
*Notes:* 

* With {acm-short} 2.12 or later, the API server URL for all types of non-{ocp-short} `ManagedCluster` resources renders automatically if you use the following importing modes:
** Entering your server URL and API token for the existing cluster. 
** Entering the `kubeconfig` file for the existing cluster. 
* The following cases can make the API server URL empty for one of the `ManagedClusters` resources:
** The non-{ocp-short} cluster is imported to the {acm-short} hub cluster before version 2.12. 
** The non-{ocp-short} cluster is manually imported to the {acm-short} hub cluster in version 2.12 through the import mode, `Run import commands`. 

[#secret-gitops]

> == _{gitops}_ token

Reviewer: not sure why this is italicized now

> When you integrate with the {gitops} operator, for every managed cluster that is bound to the {gitops} namespace through the placement and `ManagedClusterSetBinding` custom resources, a secret with a token to access the `ManagedCluster` is created in the {gitops} instance server namespace. 

Reviewer: there is also a {gitops-short} conref. After you use the full.

. Save your changes. You can now follow the GitOps workflow to manage your applications.

[#register-non-ocp]

> == Registering non-_ocp-short_ clusters to GitOps

Author: @xiangjingli let me know what you think about me making this content its own section.

Reviewer 2: Yeah, it looks good to append a new section after the existing section 
`1.2.2. Registering managed clusters to GitOps`

So the 1.2.2 is the general introduction for registering all OCP and Non OCP clusters. The new section is additional steps and checks for registering Non OCP clusters.

> You can now use the {acm-short} `GitOpsCluster` to register a non-{ocp-short} cluster to a GitOps cluster. With this capability, you can deploy application resources to the non-{ocp-short} cluster through the {gitops-short} UI. To register a non-{ocp-short} cluster to a GitOps cluster, complete the following steps:

Reviewer 3: ```suggestion
You can now use the {acm-short} `GitOpsCluster` to register a non-{ocp-short} cluster to a GitOps cluster. With this capability, you can deploy application resources to the non-{ocp-short} cluster by using the {gitops-short} UI. To register a non-{ocp-short} cluster to a GitOps cluster, complete the following steps:
```

Just removing some directional language

> . Go to the API server URL in the non-{ocp-short} `ManagedCluster` spec and validate it by running the following command: 

Author: @xiangjingli maybe this would be better for a QE reviewer, but let me know what you think about these steps. 

From reading the Jira, it seems like some of the content was more knowledge for the user rather than actions to take. Hence, I made a "Notes" section and made the steps for the users more concise. 

Let me know what you think, thanks :)

Reviewer 2: Yeah, it looks good to me

+
[source,bash]
----
oc get managedclusters eks-1

> NAME    HUB ACCEPTED   MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE

Reviewer 3: This looks like an output? I assume it's not part of the command

eks-1   true           https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
----

. If the API server URL in the non-{ocp-short} `MangedCluster` spec is empty, update it manually by completing the following steps: 
.. To complete the API server URL, edit the `MangedCluster` spec by running the following command: 
+
[source,bash]
----
oc edit managedclusters eks-1
spec:
  managedClusterClientConfigs:

>   - caBundle: ZW1wdHlDQWJ1bmRsZQo=                =====> this is a fake base64 encoded string "emptyCABundle"

Reviewer 3: Optional: We usually avoid comments in code samples. If it's a replaceable variable you can add a callout, otherwise I'd add any comments in a seperate Note tag.

Author: Thanks, Oliver, good catch! Looking more closely at it, I simply removed it, as I don't think this content is necessary. 

@xiangjingli if you thin they're particularly helpful to the user, I can add them back in, but as callouts, as Oliver suggested, thanks!

Reviewer 2: ok, it is good to simply remote it.  the caBundle content could be any valid base64 encoded string. Users won't hit any issue if they keep using our sample caBundle string.

    url: https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  ===> api server URL of the Non-OCP cluster
----

.. Save the changes then validate that the API server is completed by running the following command: 
+
[source,bash]
----
oc get managedclusters eks-1
NAME     HUB ACCEPTED  MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
eks-1    true          https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
----

. To verify that the cluster secret is generated, go to the `openshift-gitops` namespace and confirm that that `GitOpsCluster` resource status reports, `successful`. 

*Notes:* 

* With {acm-short} 2.12 or later, the API server URL for all types of non-{ocp-short} `ManagedCluster` renders automatically if you use the following importing modes:
** Entering your server URL and API token for the existing cluster. 
** Entering the `kubeconfig` file for the existing cluster. 
* The following cases can make the API server URL empty for one of the `ManagedClusters`:
** The non-{ocp-short} cluster was imported to the {acm-short} hub before version 2.12. 

> ** The non-{ocp-short} cluster was manually imported to the {acm-short} hub in version 2.12. 

Reviewer 2: The non-{ocp-short} cluster was manually imported to the {acm-short} hub in version 2.12. 
=============>
The non-{ocp-short} cluster was manually imported to the {acm-short} hub in version >=2.12 via the import mode "Run import commands manually"

The import mode "Run import commands manually" is the term used in Ui when importing clusters. We may highlight it here so that users clearly know what the manual import means.

[#secret-gitops]
== GitOps token

> When you integrate with the GitOps operator, for every managed cluster that is bound to the GitOps namespace through the placement and `ManagedClusterSetBinding` custom resources, a secret with a token to access the `ManagedCluster` is created in the namespace. The GitOps controller needs this secret to sync resources to the managed cluster. By default, the service account application manager works with the cluster admin permissions on the managed cluster to generate the GitOps cluster secret in the GitOps instance server namespace. The default namespace is `openshift-gitops`. 

Reviewer 2: a secret with a token to access the `ManagedCluster` is created in the namespace.
=============>
a secret with a token to access the `ManagedCluster` is created in the GitOps instance server namespace.

> When you integrate with the GitOps operator, for every managed cluster that is bound to the GitOps namespace through the placement and `ManagedClusterSetBinding` custom resources, a secret with a token to access the `ManagedCluster` is created in the GitOps instance server namespace. The GitOps controller needs this secret to sync resources to the managed cluster. By default, the service account application manager works with the cluster admin permissions on the managed cluster to generate the GitOps cluster secret in the GitOps instance server namespace. The default namespace is `openshift-gitops`. 

Reviewer 3: Optional: Might be worth separating since this paragraph is a little long.

. Save your changes. You can now follow the GitOps workflow to manage your applications.

[#register-non-ocp]

> == Registering non-_ocp-short_ clusters to GitOps

Reviewer: conref was missing brackets also reminder that there is a conref for GitOps ({gitops} and {gitops-short}. However, i do see in your what's new entry that you specify ACM GitOpsCluster resource. I recommend confirming with development if they are referring to OpenShift GitOps
```suggestion
== Registering non-_{ocp-short}_ clusters to {gitops}
```

Reviewer: There was also an extra plus sign that appeared in the Registering managed clusters to GitOps section
<img width="1252" alt="Screenshot 2024-10-11 at 5 21 28 PM" src="https://github.com/user-attachments/assets/40f5c64d-d9e8-4534-ab7a-1b227f9fbf0b">

You can now use the {acm-short} `GitOpsCluster` to register a non-{ocp-short} cluster to a GitOps cluster. With this capability, you can deploy application resources to the non-{ocp-short} cluster by using the {gitops-short} UI. To register a non-{ocp-short} cluster to a GitOps cluster, complete the following steps:

. Go to the API server URL in the non-{ocp-short} `ManagedCluster` spec and validate it. You should see the following output:
+
[source,bash]
----
oc get managedclusters eks-1
NAME    HUB ACCEPTED   MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
eks-1   true           https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
----

. If the API server URL in the non-{ocp-short} `MangedCluster` spec is empty, update it manually by completing the following steps: 

> .. To complete the API server URL, edit the `MangedCluster` spec by running the following command: 
> +
> [source,bash]
> ----
> oc edit managedclusters eks-1
> spec:
>   managedClusterClientConfigs:
>   - caBundle: ZW1wdHlDQWJ1bmRsZQo=          
>     url: https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com
>     cluster
> ----

Reviewer: Seems like there is a command and YAML snippet together. We usually separate these. Is the rest of the spec info intended to be presented in the command like this? 
```suggestion
.. To complete the API server URL, edit the `MangedCluster` resource `spec` by running the following command: 

+
[source,bash]
----
oc edit managedclusters eks-1
spec:
  managedClusterClientConfigs:
  - caBundle: ZW1wdHlDQWJ1bmRsZQo=          
    url: https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com
    cluster
----
```

> .. Save the changes then validate that the API server is completed by running the following command: 
> +
> [source,bash]

Reviewer: ```suggestion
.. Save the changes then validate that the API server is completed by running the following command: 

+
[source,bash]
```

----
oc get managedclusters eks-1
NAME     HUB ACCEPTED  MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
eks-1    true          https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
----

> . To verify that the cluster secret is generated, go to the `openshift-gitops` namespace and confirm that that `GitOpsCluster` resource status reports, `successful`. 
> 
> *Notes:* 
> 
> * With {acm-short} 2.12 or later, the API server URL for all types of non-{ocp-short} `ManagedCluster` renders automatically if you use the following importing modes:

Reviewer: Are the notes suppose to be aligned to step 3 or are the notes for the entire section? 
```suggestion
. To verify that the cluster secret is generated, go to the `openshift-gitops` namespace and confirm that that `GitOpsCluster` resource status reports, `successful`. 
+
*Notes:* 

* With {acm-short} 2.12 or later, the API server URL for all types of non-{ocp-short} `ManagedCluster` renders automatically if you use the following importing modes:
```

. Save your changes. You can now follow the GitOps workflow to manage your applications.

[#register-non-ocp]
== Registering non-_{ocp-short}_ clusters to {gitops}

> You can now use the {acm-short} `GitOpsCluster` to register a non-{ocp-short} cluster to a GitOps cluster. With this capability, you can deploy application resources to the non-{ocp-short} cluster by using the {gitops-short} UI. To register a non-{ocp-short} cluster to a GitOps cluster, complete the following steps:

Reviewer: every reference to GitOps should be replaced with the conref. We also do not use UI. If there are any other instances, please update it to read either console or write out user interface.
```suggestion
You can now use the {acm-short} `GitOpsCluster` to register a non-{ocp-short} cluster to a {gitops-short} cluster. With this capability, you can deploy application resources to the non-{ocp-short} cluster by using the {gitops-short} console. To register a non-{ocp-short} cluster to a {gitops-short} cluster, complete the following steps:
```

. Go to the API server URL in the non-{ocp-short} `ManagedCluster` spec and validate it. You should see the following output:
+
[source,bash]
----
oc get managedclusters eks-1
NAME    HUB ACCEPTED   MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
eks-1   true           https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 

> ----

Reviewer: ```suggestion
. Go to the API server URL in the non-{ocp-short} `ManagedCluster`  resource `spec` and validate it by running the following command.

+
[source,bash]
----
oc get managedclusters eks-1
----
+
You might receive the following output:

+
[source,bash]
----
NAME    HUB ACCEPTED   MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
eks-1   true           https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
----
```

> . If the API server URL in the non-{ocp-short} `MangedCluster` spec is empty, update it manually by completing the following steps: 

Reviewer: ```suggestion
. If the API server URL in the non-{ocp-short} `MangedCluster` resource `spec` is empty, update it manually by completing the following steps: 
```

.. To complete the API server URL, edit the `MangedCluster` resource `spec` by running the following command: 

+
[source,bash]
----
oc edit managedclusters eks-1
spec:
  managedClusterClientConfigs:

>   - caBundle: ZW1wdHlDQWJ1bmRsZQo=          

Reviewer: It seems like there is a command and an example of the YAML. This should be separated. Similar to my previous suggestion on line 66

    url: https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com
----

> .. Save the changes then validate that the API server is completed by running the following command: 
> 
> +
> [source,bash]
> ----
> oc get managedclusters eks-1
> NAME     HUB ACCEPTED  MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
> eks-1    true          https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
> ----

Reviewer: ```suggestion
.. Save the changes then validate that the API server is completed by running the following command: 

+
[source,bash]
----
oc get managedclusters eks-1
----
+
You might receive the following output:

+
[source,bash]
----
NAME     HUB ACCEPTED  MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
eks-1    true          https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
----
```

. To verify that the cluster secret is generated, go to the `openshift-gitops` namespace and confirm that that `GitOpsCluster` resource status reports, `successful`. 
+
*Notes:* 

> * With {acm-short} 2.12 or later, the API server URL for all types of non-{ocp-short} `ManagedCluster` renders automatically if you use the following importing modes:

Reviewer: ```suggestion
* With {acm-short} 2.12 or later, the API server URL for all types of non-{ocp-short} `ManagedCluster` resources renders automatically if you use the following importing modes:
```

** Entering your server URL and API token for the existing cluster. 
** Entering the `kubeconfig` file for the existing cluster. 

> * The following cases can make the API server URL empty for one of the `ManagedClusters`:

Reviewer: ```suggestion
* The following cases can make the API server URL empty for one of the `ManagedClusters` resources:
```

> ** The non-{ocp-short} cluster was imported to the {acm-short} hub before version 2.12. 

Reviewer: Why don't we use the conref for the product version? 
```suggestion
** The non-{ocp-short} cluster is imported to the {acm-short} hub cluster before version 2.12. 
```

> ** The non-{ocp-short} cluster was manually imported to the {acm-short} hub in version 2.12 through the import mode, `Run import commands`. 

Reviewer: ```suggestion
** The non-{ocp-short} cluster is manually imported to the {acm-short} hub cluster in version 2.12 through the import mode, `Run import commands`. 
```

[#secret-gitops]

> == GitOps token

Reviewer: As mentioned, each reference of GitOps should be replaced with the conref

. Save your changes. You can now follow the GitOps workflow to manage your applications.

> [#register-non-ocp]

Author: @dtthuynh or other QE reviewers, please let me know if you have concerns about me using the conref, `{gitops}`

Reviewer 4: lgtm

Reviewer: @ruici-h just a heads up, the value for the variable {gitops} is Red Hat OpenShift GitOps

. Go to the API server URL in the non-{ocp-short} `ManagedCluster` spec and validate it. You should see the following output:
+
[source,bash]
----
oc get managedclusters eks-1
NAME    HUB ACCEPTED   MANAGED CLUSTER URLS                                                      JOINED   AVAILABLE   AGE
eks-1   true           https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com  True     True        37m 
----

. If the API server URL in the non-{ocp-short} `MangedCluster` spec is empty, update it manually by completing the following steps: 
.. To complete the API server URL, edit the `MangedCluster` resource `spec` by running the following command: 

+
[source,bash]
----
oc edit managedclusters eks-1
spec:
  managedClusterClientConfigs:
  - caBundle: ZW1wdHlDQWJ1bmRsZQo=          
    url: https://5E336C922AB16684A332C10535B8D407.gr7.us-east-2.eks.amazonaws.com

>     cluster

Reviewer 2: @jc-berger  Just noticed the line 79 is not necessary. You can remove it
